### PR TITLE
[RISCV] Add Vendor Reloc and Fallback Names

### PR DESCRIPTION
**Pull request by lenary**
The trunk psABI document specifies that `R_RISCV_VENDOR` is 191, and that relocations 192-255 should be printed as `R_RISCV_CUSTOM<enum val>` in the absence of better information.

The testcase shows that these parse in `.reloc` directives, and also print correctly from both the assembler and objdump. There is almost certainly a better way to test this.

### CODE REVIEW

**Comment by jrtc27** [comment]
Is the expectation that downstream tools support `R_RISCV_CUSTOM<n>` as non-canonical aliases when they are given a defined meaning? There doesn't seem to be an ELF_RELOC_ALIAS equivalent to allow for that, so such downstream (ourselves included) would have to add extra cases to RISCVAsmBackend::getFixupKind manually. We can do that if needed, but it'd be nicer if upstream supported this more directly.

**Comment by lenary** [comment]
> Is the expectation that downstream tools support `R_RISCV_CUSTOM<n>` as non-canonical aliases when they are given a defined meaning? There doesn't seem to be an ELF_RELOC_ALIAS equivalent to allow for that, so such downstream (ourselves included) would have to add extra cases to RISCVAsmBackend::getFixupKind manually. We can do that if needed, but it'd be nicer if upstream supported this more directly.

I have a solution for `getFixupKind` to support using vendor relocation names "correctly", but I'm waiting for review on the ABI extension I've written before I can release the spec and open a PR for the corresponding implementation.

Broadly, my intention is to add another header, `ELFRelocs/RISCV_nonstandard.def`, with a `ELF_RISCV_NONSTANDARD_RELOC(vendor_symbol, name, id)` macro, and to hook that macro into `getFixupKind` so you can use the `name` directly. The reason for including the `vendor_symbol` is so that we can start to have other places use the mapping correctly. I'm not familiar with the cheri relocations, and whether they use the R_RISCV_VENDOR with a named symbol, but I took a quick look at the [corev relocs](https://github.com/openhwgroup/corev-llvm-project/blob/development/llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def) and an example from that in my scheme would be `ELF_RISCV_NONSTANDARD_RELOC(COREV, R_RISCV_CVPCREL_UI12, 224)`. The nice thing about this is it means different places in LLVM can opt into the pretty-printing (rather than the fallback) behaviour, depending on whether they want a 1:1 mapping or have more information available.

A lot more reworking of llvm-objdump will be required to support printing out the right name depending on the vendor relocation - in the meantime, it will print out `R_RISCV_CUSTOM<n>` - which I don't think is wrong, even though it could do better.

I'm trying to juggle my priorities fairly carefully, and have had to make the hard decision to prioritise the linker implementation of my specification, rather than reworking llvm-objdump, so until I can get to it (likely after 20.0) I think downstream implementations will just have to keep some modifications to `ELFRelocs/RISCV.def` if they want their name to be printed out rather than the `R_RISCV_CUSTOM<n>` name (and presumably those downstream implementations are, right now, ignoring the `R_RISCV_VENDOR` implementation when they are printing anyway).

I hope this is a reasonable/acceptable way forwards?

**Comment by lenary** [comment]
@jrtc27 ping? What do you think of the approach I outlined?

**Comment by jrtc27** [comment]
So, for CHERI-RISC-V, we predate the existence of R_RISCV_VENDOR by many years and thus just use the custom reloc encodings directly with no namespacing. Fixing that isn't worth doing because we're on the path to standardisation, which will yield the necessary relocations existing in the standard range. So your ELF_RISCV_NONSTANDARD_RELOC wouldn't help us, but might be the right shape for "proper" vendor relocations. Which may mean that it's our responsibility to deal with downstream if we're not doing things what is now the "proper" way.

**Comment by lenary** [comment]
Ok, so I guess in that case, if we can land this, downstream you would just comment out some of the `ELF_RELOC(..., ...)` that relate to relocs CHERI has its own names for already in that file. You might also need to xfail the test, but your toolchain wouldn't have a regression for users, I believe.

Once we get `ELF_RISCV_NONSTANDARD_RELOC`, the implementation I have in in `getFixupKind` ignores the vendor field for the moment anyway, and just allows the name to be used in `.reloc` directives (so vendors following the standard need a separate `.reloc ., R_RISCV_VENDOR, ..` directive - I didn't want to change how many relocation `.reloc` creates/encodes). I guess this is when you get a regression for users - because this is the point you might start to see `R_RISCV_CUSTOM<n>` instead of your nice names - but maybe you can just keep the downstream changes from the first paragraph to `ELF_RELOC(..., ...)`?

I can understand not wanting to regress downstream, and it is mostly a CHERI problem that upstream doesn't care about, but maybe with this sketched out, we can proceed on merging this patch upstream?

**Comment by jrtc27** [comment]
Yeah I've convinced myself that I have no issue with you moving ahead with this upstream. If this becomes a problem we can revisit the conversation at that later date.

File: llvm/test/MC/RISCV/custom_reloc.s [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  # CHECK-ASM-NEXT: mv a0, a0
+
+  # CHECK-OBJ: addi a0, a0, 0
+  # CHECK-OBJ-NEXT: R_RISCV_VENDOR    VENDOR_NAME
+  # CHECK-OBJ-NEXT: R_RISCV_CUSTOM192 my_foo+0x1
+
+  nop
+  # CHECK-ASM: nop
+  # CHECK-OBJ: addi zero, zero, 0x0
+
```
Comment (MaskRay) [inline llvm/test/MC/RISCV/custom_reloc.s:42]: delete trailing blank lines

File: llvm/test/MC/RISCV/custom_reloc.s [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# RUN: llvm-mc -triple riscv32 < %s \
+# RUN:   | FileCheck -check-prefix=CHECK-ASM %s
+# RUN: llvm-mc -triple riscv64 < %s \
+# RUN:   | FileCheck -check-prefix=CHECK-ASM %s
+
+# RUN: llvm-mc -filetype=obj -triple riscv32 < %s \
```
Comment (MaskRay) [inline llvm/test/MC/RISCV/custom_reloc.s:6]: `< %s` => `%s`

File: llvm/test/MC/RISCV/custom_reloc.s [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+# RUN: llvm-mc -filetype=obj -triple riscv64 < %s \
+# RUN:   | llvm-objdump -dr -M no-aliases  - \
+# RUN:   | FileCheck -check-prefix=CHECK-OBJ %s
+
+  # CHECK-ASM: .text
+
+  # If `.set` is used instead, the relocation expression will be evaluated.
+  # This leaves an undef, global symbol called VENDOR_NAME, which I'm not happy
+  # about.
+  .global VENDOR_NAME
```
Comment (MaskRay) [inline llvm/test/MC/RISCV/custom_reloc.s:18]: you can drop this `.globl`. Symbols only referenced by `.reloc` directives are emitted to .symtab as undefined symbols.

File: llvm/test/MC/RISCV/custom_reloc.s (hunk 2 above)
Comment (lenary) [reply llvm/test/MC/RISCV/custom_reloc.s]: Done

File: llvm/test/MC/RISCV/custom_reloc.s (hunk 3 above)
Comment (lenary) [reply llvm/test/MC/RISCV/custom_reloc.s]: Done

File: llvm/test/MC/RISCV/custom_reloc.s (hunk 1 above)
Comment (lenary) [reply llvm/test/MC/RISCV/custom_reloc.s]: Done

**Review by lenary — COMMENTED** [review]
Sorry for the force push, I have some branches stacked on this in my local repo which I wanted to update too.

**Comment by lenary** [comment]
ping?

**Review by MaskRay — APPROVED** [review]
if jrtc27 is happy

**Pull request merged**